src/service/visibility-impl.js

```diff
@@ -0,0 +1,142 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {getService} from '../service';
+import {resourcesFor} from '../resources';
+import {viewportFor} from '../viewport';
+
+/**
+ * This type signifies a callback that gets called when visibility conditions
+ * are met.
+ * @typedef {function()}
+ */
+let VisibilityListenerCallbackDef;
+
+/**
+ * @typedef {Object<string, JSONObject|VisibilityListenerCallbackDef>}
+ */
+let VisibilityListenerDef;
+
+/**
+ * Allows tracking of AMP elements in the viewport.
+ *
+ * This class allows a caller to specify conditions to evaluate when an element
+ * is in viewport and for how long. If the conditions are satisfied, a provided
+ * callback is called.
+ */
+export class Visibility {
+
+  /** @param {!Window} */
+  constructor(win) {
+    this.win_ = win;
+
+    /**
+     * key: resource id.
+     * value: [{ config: <config>, callback: <callback>}]
+     * @type {Object<string, Array.<VisibilityListenerDef>>}
+     * @private
+     */
+    this.listeners_ = Object.create(null);
+
+    /** @private {Array<!Resource>} */
+    this.resources_ = [];
+
+    /** @private @const {function} */
+    this.boundScrollListener_ = this.scrollListener_.bind(this);
+
+    /** @private {boolean} */
+    this.scrollListenerRegistered_ = false;
+
+    /** @private {!Resources} */
+    this.resourcesService_ = resourcesFor(this.win_);
+  }
+
+  /** @private */
+  registerForViewportEvents_() {
+    if (!this.scrollListenerRegistered__) {
+      const viewport = viewportFor(this.win_);
+
+      // Currently unlistens are not being used. In the event that no resources
+      // are actively being monitored, the scrollListener should be very cheap.
+      viewport.onScroll(this.boundScrollListener_);
+      viewport.onChanged(this.boundScrollListener_);
+      this.scrollListenerRegistered_ = true;
+    }
+
+  }
+
+  /**
+   * @param {!JSONObject} config
+   * @param {!VisibilityListenerCallbackDef} callback
+   */
+  listenOnce(config, callback) {
+    const element = this.win_.document.getElementById(config['selector']
+        .slice(1));
+    const res = this.resourcesService_.getResourceForElement(element);
+    const resId = res.getId();
+
+    this.registerForViewportEvents_();
+
+    this.listeners_[resId()] = (this.listeners_[resId()] || []);
+    this.listeners_[resId()].push({
+      config: config,
+      callback: callback,
+    });
+    this.resources_.push(res);
+
+    this.scrollListener_(); // Should this be async?
```

Comment:
> Should this be async?

Might be good. Registering a delay, then canceling if there is a registered delay in `#scrollListener_` should work well.